Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+            ],
+
+            (Pending pending, RemoveProduct((var productId, var toRemove, var price) pricedProductItem)) =>
+                pending.ProductItems.TryGetValue($"{productId}{price}", out var quantity) && quantity >= toRemove
+                    ?
+                    [
+                        new ProductRemoved(pricedProductItem, Now)
+                    ]
+                    : throw new InvalidOperationException("Not enough Products!"),
+
+            (Pending, Confirm(var clientId)) =>
+            [
+                new Confirmed(clientId, Now)
+            ],
+
+            (Pending, Cancel) =>
+            [
+                new Cancelled(Now)
+            ],
+
+            (Closed, Confirm) => [],
+
+            (Closed, Cancel) => [],
+
+            _ => throw new InvalidOperationException(nameof(command))
+        };
+
+    private DateTimeOffset Now => timeProvider.GetLocalNow();
+}
+
+public class ShoppingCartModel
+{
+    public required Guid Id { get; init; }
+    public Guid? ClientId { get; set; }
+    public required ShoppingCartStatus Status { get; set; }
+    public List<PricedProductItem> ProductItems { get; init; } = [];
+    public required DateTimeOffset OpenedAt { get; set; }
+    public DateTimeOffset? ConfirmedAt { get; set; }
+    public DateTimeOffset? CanceledAt { get; set; }
+
+    public class PricedProductItem
+    {
+        public required Guid ProductId { get; init; }
+        public required int Quantity { get; set; }
+        public required decimal Price { get; init; }
+        public required DateTimeOffset AddedAt { get; set; }
+        public DateTimeOffset? UpdatedAt { get; set; }
+    }
+
+    public enum ShoppingCartStatus
+    {
+        Pending,
+        Confirmed
+    }
+}
+
+public interface IShoppingCartRepository
+{
+    ValueTask<ShoppingCartModel?> Find(Guid id, CancellationToken ct);
+
+    Task Store(Guid id, ShoppingCartModel? current, ShoppingCartEvent[] events, CancellationToken ct);
+
+    async Task GetAndStore(Guid id, Func<ShoppingCartModel?, ShoppingCartEvent[]> handle, CancellationToken ct)
+    {
+        var state = await Find(id, ct);
+
+        var events = handle(state);
+
+        await Store(id, state, events, ct);
+    }
+}
+
+public class ShoppingClassCommandHandler(
+    ShoppingClassDecider decider,
+    IShoppingCartRepository repository
+)
+{
+    public Task Handle(Guid id, ShoppingCartCommand command, CancellationToken ct) =>
+        repository.GetAndStore(
+            id,
+            model => decider.Decide(command, Map(model)),
+            ct
+        );
+
+    private ShoppingCart Map(ShoppingCartModel? model)
+    {
+        if (model == null)
+            return new Empty();
+
+        switch (model.Status)
+        {
+            case ShoppingCartModel.ShoppingCartStatus.Pending:
+                return new Pending(
+                    model.ProductItems.ToImmutableDictionary(ks => $"{ks.ProductId}{ks.Price}", vs => vs.Quantity)
```

Comment:
ew, string'd price as a key?
can end up with dups keys and all sorts of ugliness